File: benchmarks/remediation/fixtures/js-path-sendfile/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the app.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
//
// The response object is a recorder, so no file is sent. Two routes in the one file are
// vulnerable, so repairing only one of them leaves the exploit check failing.
const assert = require('node:assert/strict');
const path = require('node:path');

const ATTACK = '../../../etc/passwd';
const LEGITIMATE = 'logo.png';
const BASES = { sendAsset: '/srv/assets', sendThumbnail: '/srv/assets/thumbs' };

function response() {
  const record = { sent: [], status: null };
  record.res = {
    sendFile: (target) => { record.sent.push(String(target)); return 'sent'; },
    status: (code) => { record.status = code; return { end: () => 'ended' }; },
  };
  return record;
}

function loadLocal() {
  return require(path.resolve(__dirname, '..', 'app.js'));
}

function call(module, name, value) {
  const record = response();
  const request = { params: { name: value }, query: { name: value } };
  module[name](request, record.res);
  return record;
}

function exploit() {
  const module = loadLocal();
  const escaped = [];
  for (const name of Object.keys(BASES)) {
    for (const sent of call(module, name, ATTACK).sent) {
      if (!path.resolve(sent).startsWith(`${BASES[name]}${path.sep}`)) escaped.push(sent);
    }
  }
  if (escaped.length > 0) {
    process.stderr.write(`vulnerability present: ${escaped.length} route(s) send a path outside their directory\n`);
    process.exit(1);
  }
  process.exit(0);
}

function behavior() {
  const module = loadLocal();
  for (const name of Object.keys(BASES)) {
    const record = call(module, name, LEGITIMATE);
    assert.equal(record.status, null);
    assert.deepEqual(record.sent, [path.join(BASES[name], LEGITIMATE)]);
  }
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  const original = require(originalPath);
  assert.deepEqual(call(original, 'sendAsset', ATTACK).sent, ['/etc/passwd']);
  assert.deepEqual(call(original, 'sendThumbnail', ATTACK).sent, ['/etc/passwd']);
  const repaired = require(repairedPath);
  for (const name of Object.keys(BASES)) {
    const blocked = call(repaired, name, ATTACK);
    assert.deepEqual(blocked.sent, []);
    assert.equal(blocked.status, 400);
    const allowed = call(repaired, name, LEGITIMATE);
    assert.deepEqual(allowed.sent, [path.join(BASES[name], LEGITIMATE)]);
  }
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);

File: benchmarks/remediation/fixtures/python-path-send-file/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.

Flask is not installed in the sandbox, so the module is replaced by recording doubles before
the module under test is imported. send_file records the path it was handed; it opens nothing.
"""
import importlib.util
import json
import os
import sys
import types

ATTACK = "../../etc/passwd"
LEGITIMATE = "2026/march.pdf"
BASE = "/srv/reports"
SENT = []
ROUTES = {}


class HTTPError(Exception):
    def __init__(self, code):
        Exception.__init__(self, str(code))
        self.code = code


class FakeRequest:
    def __init__(self):
        self.args = {}


REQUEST = FakeRequest()


def route(rule, **kwargs):
    def decorate(view):
        ROUTES[rule] = view
        return view

    return decorate


class FakeFlask:
    def __init__(self, name):
        self.name = name

    route = staticmethod(route)


def send_file(target, **kwargs):
    SENT.append(str(target))
    return b"%PDF-1.4 fixture"


flask = types.ModuleType("flask")
flask.Flask = FakeFlask
flask.request = REQUEST
flask.abort = lambda code, *a, **k: (_ for _ in ()).throw(HTTPError(code))
flask.send_file = send_file
sys.modules.setdefault("flask", flask)


def load(path, name="app_under_test"):
    ROUTES.clear()
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def fetch(name):
    del SENT[:]
    REQUEST.args = {"name": name}
    try:
        return ROUTES["/reports"](), None
    except HTTPError as error:
        return None, error.code


def exploit():
    load(local_path())
    fetch(ATTACK)
    escaped = [item for item in SENT if not os.path.normpath(item).startswith(BASE + os.sep)]
    if escaped:
        sys.stderr.write("vulnerability present: the resolved path escapes the report directory\n")
        sys.exit(1)
    sys.exit(0)


def behavior():
    load(local_path())
    body, code = fetch(LEGITIMATE)
    assert code is None and body is not None
    assert SENT and os.path.normpath(SENT[0]) == os.path.join(BASE, LEGITIMATE)
    sys.exit(0)


def compare(original_path, repaired_path):
    load(original_path, "original")
    fetch(ATTACK)
    assert os.path.normpath(SENT[0]) == "/etc/passwd"
    load(repaired_path, "repaired")
    body, code = fetch(ATTACK)
    assert SENT == [] and code == 400
    body, code = fetch(LEGITIMATE)
    assert code is None and os.path.normpath(SENT[0]) == os.path.join(BASE, LEGITIMATE)
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])
